### PR TITLE
make encapsulate work with isolated environments

### PR DESCRIPTION
**Pull request by isidentical**

### CODE REVIEW

File: adapter/src/dbt/adapters/fal/utils.py [hunk 1]
```diff
@@ -65,3 +68,16 @@ def fetch_environment(project_root: str, environment_name: str) -> Tuple[BaseEnv
...
     return environments[environment_name], False
+
+
+def db_adapter_config(config: RuntimeConfig) -> RuntimeConfig:
+    """Return a config object that has the database adapter as its primary. Only
+    applicable when the underlying db adapter is encapsulated."""
+    if hasattr(config, "sql_adapter_credentials"):
+        python_adapter_credentials = config.credentials
+        new_config = replace(config, credentials=config.sql_adapter_credentials)
+        new_config.python_adapter_credentials = config.credentials
```
Comment (chamini2) [inline adapter/src/dbt/adapters/fal/utils.py:79]: TIL about replace, but does this make sense:

```suggestion
        new_config = replace(config, credentials=config.sql_adapter_credentials, python_adapter_credentials=config.credentials)
```

?

File: adapter/src/dbt/adapters/fal/utils.py [hunk 2]
```diff
@@ -65,3 +68,16 @@ def fetch_environment(project_root: str, environment_name: str) -> Tuple[BaseEnv
         )
 
     return environments[environment_name], False
+
+
+def db_adapter_config(config: RuntimeConfig) -> RuntimeConfig:
+    """Return a config object that has the database adapter as its primary. Only
+    applicable when the underlying db adapter is encapsulated."""
```
Comment (chamini2) [inline adapter/src/dbt/adapters/fal/utils.py:75]: great approach

**Review by chamini2 — APPROVED** [review]
Thanks for getting this to work! Check a few coments

**Comment by isidentical** [comment]
This should be good to go, thanks guys I'll try to look into the reviews @chamini2 @mederka

**Review by turbo1912 — APPROVED** [review]

File: adapter/src/dbt/adapters/fal/utils.py (hunk 1 above)
Comment (isidentical) [reply adapter/src/dbt/adapters/fal/utils.py:79]: AFAIK `python_adapter_credentials` is just an attribute we attach at the runtime, not something that is on the config (might be wrong on this) of the official dbt-core==1.3. `replace()` only works for standard parameters:
```
>>> from dataclasses import dataclass, replace
>>> 
>>> @dataclass
... class T:
...     name: str
... 
>>> 
>>> t = T(name="hey")
>>> replace(t, name="x")
T(name='x')
>>> replace(t, name="x", random_attr="y")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/seaworth/.pyenv/versions/3.9.13/lib/python3.9/dataclasses.py", line 1284, in replace
    return obj.__class__(**changes)
TypeError: __init__() got an unexpected keyword argument 'random_attr'
```

**Pull request merged**